### ISSUE TITLE
Renaming test files isn't handled properly

### ISSUE THREAD

**Issue by d-albrecht** [scalameta/metals#4863]
**Describe the bug**

Project structure: Pure Scala project with ScalaTest (AnyWordSpec-style) tests managed by sbt opened in VSCodium using metals
Renaming a test file from the VSCode Explorer view doesn't fully propagate to the Testing view because (most likely) Metals isn't properly handling the new file name.

**To Reproduce** Steps to reproduce the behavior:

1. Create a test file with some test class and test cases.
2. See that the test cases appear in the "Testing" view and you can both run them from there and click "Go to Test" to open the file.
3. Go to the "Explorer" view and rename the file (not the class).
4. See that the test cases still appear in the "Testing" view (and can be run) but you can't go to the sources anymore (it leads you to an empty "file doesn't exist" page).
5. Reopening VSCode solves the problem.

Alternatively:

5. Opening the renamed file from the Explorer view and also renaming the test **class** fixes the issue.
6. But, going into the file and renaming the test **class** back to the former name again breaks the "go to" linkage. Probably due to Metals caching test location data somewhere and it recognizing the "old" name.

**Expected behavior**

Metals should be able to adopt to renamed files immediately.

**Screenshots**

None

**Installation:**

- Operating system: Linux (Fedora)
- VSCode version: 1.74.2
- VSCode extension version: 1.21.0
- Metals version: (found in VSCode settings, under `metals.serverVersion`): 0.11.10

**Additional context**

None

**Search terms**

None

**Comment by tgodzik** [comment]
Thanks for reporting! Looks like a solid reproduction to follow up on.

**Comment by d-albrecht** [comment]
Just to improve my own understanding of the (internal) process here, what was the reason/effect of moving the issue from metals-vscode to here (metals)? Is this just the correct place for not yet assigned/evaluated bugs? Or have you tried this in other IDEs (with their respective metals plug-in) and have discovered that this is a general issue with all plug-ins?

I don't use other metals plug-ins other than the VSCode one and don't want to install them just for testing, so I assumed that this might be VSCode specific. Would it make any difference if I had issued this bug report in this repo initially? Or is this something I shouldn't worry too much about and the one in charge of triaging all bug reports will sort them into the right buckets eventually anyway.

**Comment by tgodzik** [comment]
We try to keep most of the logic inside the Metals language server (which is this repo) and `metals-vscode` is an extension that starts the server and adds some additional functionality to VS Code. Test explorer is mostly handled on the server side, which is why I decided to move it here.

Most issues will be related to the server with only a minor fraction to the VS Code extension.

**Comment by kpodsiad** [comment]
This issue is tricky due to the fact what is our source of truth and how we store found tests. We're using https://build-server-protocol.github.io/docs/extensions/scala#scala-test-classes-request to obtain a set of fully qualified class names (FCQN) which is then transformed to the form which is suitable for the Test Explorer view. Adding or removing tests is based on diffing obtained FCQNs with the current one in use, that's how we handle addition or removal of tests. Running tests is also based on FQCN, that's why it works.

To sum up, there is no low hanging fix for this, but I'll try to think about what we can do.

**Comment by d-albrecht** [comment]
That sort of ties into the fact that at times new test cases don't appear in the testing view and also that new test cases don't get the green/red triangle/run or checkmark/cross in front of the test case to run this (singular) case from the source file. I haven't yet found a good way to post them, so I will just mention that here.

I haven't yet read the whole linked documentation but I don't know if that would tell me anything useful. But that all combined, it might be a viable idea (if my understanding adds up) to replace the tree view from time to time after bigger changes if diffing isn't working all the time. Again, just an idea, don't know if this would help.

Another minor question in this context: all options to run multiple test cases at once is limited to a) whole projects, b) whole packages, and c) whole test **classes**. But in the context of "AnyWordSpec", I'd like to run all tests that use the same "should" name (``"Some class" should {...}``). Because I can put several "should"s into one class. Is there any intentions/chances to enable this?

**Comment by tgodzik** [comment]
>  But in the context of "AnyWordSpec", I'd like to run all tests that use the same "should" name ("Some class" should {...}). Because I can put several "should"s into one class. Is there any intentions/chances to enable this?

That would need to be a feature request, but I think it's perfectly doable within the logic we have here https://github.com/scalameta/metals/blob/main/metals/src/main/scala/scala/meta/internal/metals/testProvider/frameworks/ScalatestTestFinder.scala

**Comment by tgodzik** [comment]
> That sort of ties into the fact that at times new test cases don't appear in the testing view and also that new test cases don't get the green/red triangle/run or checkmark/cross in front of the test case to run this (singular) case from the source file. I haven't yet found a good way to post them, so I will just mention that here.

There might some cases to cover, it seems this might sometimes be flaky with a lot of changes at the same time.

**Comment by kpodsiad** [comment]
@d-albrecht can you provide a whole scalatest test file with a few examples of tests? I'm not a Scala user so having concrete example would allow me to implement and test exactly your feature request, when I'd find some spare time ;)

**Comment by d-albrecht** [comment]
Sure, no problem. These two files should represent a minimum working test project.
I did them on the fly, but quickly compile-checked them, so, while not the most elegant/useful, these should do the trick.
If you need more test cases, just copy paste the individual test cases into the "should"-scope, but you have to rename/reword the test-name (the "case x: ..." thing) because the test runner can't handle two test cases that have the exact same name (just increment the "case x" part).

build.sbt
```
ThisBuild / scalaVersion := "3.2.2"
ThisBuild / organization := "test"
lazy val commonSettings = Seq(
	libraryDependencies += "org.scalactic" %% "scalactic" % "3.2.12",
	libraryDependencies += "org.scalatest" %% "scalatest" % "3.2.12" % Test,
	libraryDependencies += "org.scalatestplus" %% "junit-4-13" % "3.2.12.0" % "test",
	libraryDependencies += "junit" % "junit" % "4.13.2" % Test,
)
lazy val test = (project in file("."))
	.settings(commonSettings)
```
src/test/scala/Test.scala
```
import org.junit.runner.RunWith
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec
import org.scalatestplus.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class Test extends AnyWordSpec with Matchers {
	"This test" should {
		"case 1: be able to run" in {
			5 + 5 should be (10)
		}
		"case 2: fail" in {
			assertThrows[IndexOutOfBoundsException] {
				List()(5)
			}
		}
	}
}
```
